src/resources.rs

```diff
@@ -0,0 +1,369 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Structs to manage device resources.
+//!
+//! The high level flow of resource management among the VMM, the device manager, and the device
+//! is as below:
+//! 1) the VMM creates a new device object.
+//! 2) the VMM asks the new device object for its resource constraints.
+//! 3) the VMM allocates resources for the device object according to resource constraints.
+//! 4) the VMM passes the allocated resources to the device object.
+//! 5) the VMM registers the new device onto corresponding device managers according the allocated
+//!    resources.
+
+use std::{u16, u32, u64};
+
+/// Enumeration describing a device's resource constraints.
+pub enum ResourceConstraint {
+    /// Constraint for an IO Port address range.
+    PioAddress {
+        /// Allocate resource within the range [`min`, `max`] if `min` is not u16::MAX.
+        min: u16,
+        /// Allocate resource within the range [`min`, `max`] if `max` is not zero.
+        max: u16,
```

Comment:
Using these 'special values' is error prone, especially when they are in public interfaces. Since they can only be discovered from the comments or even from the code, It's easy for people to forget to set or check that values and that can not be caught by compiler. So I still think Option is a good choice here. Since this is only for ResourceConstrait (not for Resource), I don't see lot of coding inconvenience on checking Some() or using unwrap().

Reply:
Great idea, need to think in rust way:)